core/src/main/java/com/linecorp/armeria/common/DeferredHttpResponse.java

```diff
@@ -50,9 +70,28 @@ void delegate(HttpResponse delegate) {
 
     void delegateWhenComplete(CompletionStage<? extends HttpResponse> stage) {
         requireNonNull(stage, "stage");
+
+        // Propagate exception to the upstream future.
+        whenComplete().handle((unused, cause) -> {
+            final CompletableFuture<? extends HttpResponse> future = stage.toCompletableFuture();
+            if (cause != null && !future.isDone()) {
+                if (MONO_TO_FUTURE_CLASS != null && MONO_TO_FUTURE_CLASS.isAssignableFrom(future.getClass())) {
+                    // A workaround for 'MonoToCompletableFuture' not propagating cancellation to the upstream
+                    // publisher when it completes exceptionally.
+                    future.cancel(true);
+                } else {
+                    future.completeExceptionally(cause);
+                }
+            }
+            return null;
+        });
```

Comment:
How about extracting this method as an internal utility and also use it for others such as `ResponseConversionUtil`, `cancelOnDownstreamException()` and so on?
```java
class CompletableFutureUtil {
   
    static void propagateException(CompletableFuture<?> source, CompletionStage<?> target) {
        source.handle((unused, cause)) -> {
             ....
        })
    }
}

CompletableFutureUtil.propagateException(whenComplete(), stage);
```

Reply:
Let me try! 😆